src/datanode/src/sql/insert.rs

```diff
@@ -123,6 +139,153 @@ impl SqlHandler {
             region_number: 0,
         }))
     }
+
+    fn build_request_from_batch(
+        stmt: Insert,
+        table: TableRef,
+        batch: RecordBatch,
+        query_ctx: QueryContextRef,
+    ) -> Result<SqlRequest> {
+        let (catalog_name, schema_name, table_name) =
+            table_idents_to_full_name(stmt.table_name(), query_ctx)?;
+
+        let columns: Vec<String> = stmt.columns().iter().map(|c| (*c).clone()).collect();
+        let schema = table.schema();
+        let columns = if columns.is_empty() {
+            schema
+                .column_schemas()
+                .iter()
+                .map(|c| c.name.to_string())
+                .collect()
+        } else {
+            columns
+        };
+        let columns_num = columns.len();
+
+        ensure!(
+            batch.num_columns() == columns_num,
+            ColumnValuesNumberMismatchSnafu {
+                columns: columns_num,
+                values: batch.num_columns(),
+            }
+        );
+
+        let batch_schema = &batch.schema;
+        let batch_columns = batch_schema.column_schemas();
+        assert_eq!(batch_columns.len(), columns_num);
+        let mut columns_values = HashMap::with_capacity(columns_num);
+
+        for (i, column_name) in columns.into_iter().enumerate() {
+            let column_schema = schema
+                .column_schema_by_name(&column_name)
+                .with_context(|| ColumnNotFoundSnafu {
+                    table_name: &table_name,
+                    column_name: &column_name,
+                })?;
+            let expect_datatype = column_schema.data_type.as_arrow_type();
+            let batch_datatype = batch_columns[i].data_type.as_arrow_type();
```

Comment:
Seems you are relying on the sameness ordering of stmt columns and recordbatch columns? What if `insert into monitor2(ts, host,memory) select host,ts,memory from monitor;`, is it legal?

Reply:
Then this statement would run fail. It's the user's responsibility to keep the columns order the same between `insert` and `select`.